spring-kafka/src/test/java/com/baeldung/partitioningstrategy/KafkaApplicationUnitTesting.java

```diff
@@ -0,0 +1,164 @@
+                    expectedPartition);
+            }
+        });
+    }
+
+    @Test
+    public void givenProducerWithSameKeyMessages_whenSendingMessages_shouldReceiveInProducedOrder() throws InterruptedException {
+        kafkaProducer.send("order-topic", "partitionA", "message1");
+        kafkaProducer.send("order-topic", "partitionA", "message3");
+        kafkaProducer.send("order-topic", "partitionA", "message4");
+
+        Thread.sleep(1000);
+
+        List<ReceivedMessage> records = kafkaMessageConsumer.getReceivedMessages();
+
+        StringBuilder resultMessage = new StringBuilder();
+        records.forEach(record -> resultMessage.append(record.getMessage()));
+        String expectedMessage = "message1message3message4";
+
+        assertEquals(
+            "Messages with the same key should be received in the order they were produced within a partition",
+            expectedMessage,
+            resultMessage.toString()
+        );
+    }
+
+    @Test
+    public void givenCustomPartitioner_whenSendingMessages_shouldRouteToCorrectPartition() throws InterruptedException {
+        // Configure the producer with the custom partitioner
+        KafkaTemplate<String, String> kafkaTemplate = setProducerToUseCustomPartitioner();
+
+        kafkaTemplate.send("order-topic", "123_premium", "Order 123, Premium order message");
+        kafkaTemplate.send("order-topic", "456_normal", "Normal order message");
+
+        Thread.sleep(1000);
+
+        List<ReceivedMessage> records = kafkaMessageConsumer.getReceivedMessages();
+
+        // Validate that messages are routed to the correct partition based on customer type
+        for (ReceivedMessage record : records) {
+            if ("123_premium".equals(record.getKey())) {
+                assertEquals("Premium order message should be in partition 0", 0, record.getPartition());
+            } else if ("456_normal".equals(record.getKey())) {
+                assertEquals("Normal order message should be in partition 1", 1, record.getPartition());
+            }
+        }
+    }
+
+    @Test
+    public void givenCustomPartitioner_whenSendingMessages_shouldConsumeOnlyFromSpecificPartition() throws InterruptedException {
+        KafkaTemplate<String, String> kafkaTemplate = setProducerToUseCustomPartitioner();
+
+        kafkaTemplate.send("order-topic", "123_premium", "Order 123, Premium order message");
+        kafkaTemplate.send("order-topic", "456_normal", "Normal order message");
+
+        Thread.sleep(1000);
+
+        consumer.assign(Collections.singletonList(new TopicPartition("order-topic", 0)));
+        ConsumerRecords<String, String> records = consumer.poll(Duration.ofMillis(100));
+
+        for (ConsumerRecord<String, String> record : records) {
+            assertEquals("Premium order message should be in partition 0", 0, record.partition());
+            assertEquals("123_premium", record.key());
+        }
+
+        consumer.close();
```

Comment:
This close should be done after every test in an `@AfterEeach` - or it might actually already be handled by the `@SpringBootTest` - you can set a breakpoint inside the close method and see if it is already called per test

Reply:
I've checked, it actually handled by the @SpringBootTest, therefore I've removed it.